File: ws/bonilla/U3/WS35Observer/Observerjs/model/Stock.js
class Stock {
    constructor() {
        this.symbol = undefined;
        this.price = undefined;
        this.investors = [];
    }

    addObserver(iinvestor) {
        this.investors.push(iinvestor);
    }

    deleteObserver(iinvestor) {
        this.investors = this.investors.filter(i => i !== iinvestor);
    }

    notifyObservers(args) {
        for (const investor of this.investors) {
            investor.update(this, args);
        }
    }

    getSymbol() {
        return this.symbol;
    }

    setSymbol(symbol) {
        this.symbol = symbol;
    }

    getPrice() {
        return this.price;
    }

    setPrice(price) {
        this.price = price;
    }

    getInvestors() {
        return this.investors;
    }

    setInvestors(investors) {
        this.investors = investors;
    }
}

module.exports = Stock;

File: ws/fuentes/U3/WS35OBSERVER/OBSERVERPY/ec/edu/espe/saleApp/model/stock.py
from typing import List, Any
from abc import ABC, abstractmethod

class Stock:
    def __init__(self):
        self._symbol: str = ""
        self._price: float = 0.0
        self._investors: List[IInvestor] = []

    def add_observer(self, investor):
        self._investors.append(investor)

    def delete_observer(self, investor):
        self._investors.remove(investor)

    def notify_observers(self, arg: Any):
        for investor in self._investors:
            investor.update(self, arg)

    def get_symbol(self):
        return self._symbol

    def set_symbol(self, symbol: str):
        self._symbol = symbol
        self.notify_observers(symbol)

    def get_price(self):
        return self._price

    def set_price(self, price: float):
        self._price = price
        self.notify_observers(price)


class IInvestor(ABC):
    @abstractmethod
    def update(self, stock: Stock, arg: Any):
        pass
